### PR TITLE
skip upload if file exists

### PR DESCRIPTION
**Pull request by fraserxu**
This pull request trying to fix the deployment issue we have while syncing assets to production s3 bucket.

Currently it's failing because we have conflict while uploading and conflict is not handled at the moment.

This fix simply skip the upload if there's conflict instead of raise an error.

### CODE REVIEW

File: lib/static_sync/processor.rb [hunk 1]
```diff
@@ -46,6 +46,8 @@ def handle_conflict(file)
         if file.cached?
           raise ConflictError, "modifications to existing cached files are not allowed."
```
Comment (nemski) [inline lib/static_sync/processor.rb]: If we are reaching this exception, how does this change resolve that?

Comment (fraserxu) [inline lib/static_sync/processor.rb]: We are not using that api in the code at the moment. 

The assets we are building on production(this is true for the webpack assets at least for now) is based on the hash value of the content, if cache exists that actually means it's the same file, it's not override.

I'll work on the sprockets assets after the work for webpack is done.

Comment (nemski) [inline lib/static_sync/processor.rb]: But we are raising the exception here:

```
StaticSync::Processor::ConflictError: modifications to existing cached files are not allowed.
...bundler/gems/static_sync-f982792bb3f5/lib/static_sync/processor.rb:47:in `handle_conflict'
```

So somehow https://github.com/envato/static_sync/pull/1/files#diff-1b2da3445b595915666a129a9f8d5c6dL45 is matching.

Comment (nemski) [inline lib/static_sync/processor.rb]: It's matching because we are setting this flag https://github.com/compactcode/static_sync/blob/1b94905c4ddd12bc0d12dac9e4e060317e5bd277/lib/static_sync/config.rb#L43

Comment (fraserxu) [inline lib/static_sync/processor.rb]: But we are using `'conflict_mode' => 'fail_if_cached'` in the code instead of `@config.fail_on_conflict_if_cached?`, is it the same thing? It looks confusing.

Comment (taoza) [inline lib/static_sync/processor.rb]: @fraserxu @nemski Maybe we should change this config: https://github.com/envato/marketplace/blob/master/lib/tasks/assets.rake#L102

File: lib/static_sync/processor.rb [hunk 2]
```diff
@@ -12,6 +12,7 @@ class ConflictError < StandardError
     def initialize(config, storage = nil)
       @config  = config
       @storage = storage || StaticSync::Storage.new(config)
+      @skip    = false
```
Comment (taoza) [inline lib/static_sync/processor.rb]: Should `@skip` to be scoped by the file or the processor? The reason I ask is because `@skip` is set in `#handle_conflict` which seems to be scoped by files.

Comment (nemski) [inline lib/static_sync/processor.rb]: Haha, yes it should be!

File: lib/static_sync/processor.rb [hunk 3]
```diff
@@ -46,9 +47,16 @@ def handle_conflict(file)
         if file.cached?
           raise ConflictError, "modifications to existing cached files are not allowed."
         end
+      elsif @config.ignore_conflict?
+        @skip << file
+        log.info("#{file} already exist, skipping.") if @config.log
       end
     end
 
+    def skip(file)
```
Comment (johnsyweb) [inline lib/static_sync/processor.rb]: Perhaps `#skip?` may be a better name for this predicate?

File: lib/static_sync/processor.rb [hunk 4]
```diff
@@ -27,7 +28,7 @@ def sync
               log.info("Uploading #{file}") if @config.log
             end
             begin
-              @storage.create(current_file.headers)
+              @storage.create(current_file.headers) unless skip(file)
```
Comment (johnsyweb) [inline lib/static_sync/processor.rb]: Yet another branch in this (already complex) method makes it hard to follow the possible outcomes. How can we simplify this?

**Comment by taoza** [comment]
Potentially:
1. `s/@skip/@skip_files/`
2. `s/def skip/def skip?/`

But this should work too. :+1:

**Comment by johnsyweb** [comment]
How do we test this change?

File: lib/static_sync/processor.rb (hunk 4 above)
Comment (nogates) [inline lib/static_sync/processor.rb]: quick solution:

```ruby
def sync 
  log_sync
  each_local_filter_file do |file|
    next if @storage.has_version?(current_file.version)
    store_file(file)
  end
end

def store_file(file)
  uploadable_file(file).tap do |current_file|
    @storage.create(current_file.headers) unless skip(file)
  end
rescue => error 
  log.error("Failed to upload #{file}")
  raise error
end

def uploadable_file(file)
 Uploadable.new(file, @config).tap do |current_file|
   if @storage.has_file?(current_file.version)
     handle_conflict(current_file)
   else
     log.info("Uploading #{ file }") if @config.log
   end
  end
end

def log_sync
  return unless @config.log
  log.info("Synching #{@config.local_directory} to #{@config.remote_directory}.")
end

def each_local_filter_file
  Dir.chdir(@config.local_directory) do
    local_filtered_files.each do |file|
      yield file
    end
  end
end
```

**Comment by nemski** [comment]
@johnsyweb the tests are broken on master currently :(

File: lib/static_sync/processor.rb (hunk 4 above)
Comment (nemski) [inline lib/static_sync/processor.rb]: @nogates can you PR that to this PR? My mental merge is not working today.

Comment (nogates) [inline lib/static_sync/processor.rb]: Sure thing!

**Comment by nogates** [comment]
> @johnsyweb the tests are broken on master currently :(

I fixed those in my brach: https://github.com/envato/static_sync/pull/2

Also, added a dumb test for this. @johnsyweb @taoza @fraserxu @nemski

**Comment by johnsyweb** [comment]
Great work. We should consider sending this upstream.
